torchx/schedulers/ray/ray_driver.py

```diff
@@ -0,0 +1,144 @@
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import contextlib
+import importlib
+import json
+import logging
+import os
+import sys
+from typing import Dict, List, Optional
+
+import ray
+from ray.train.utils import get_address_and_port
+from ray.util.placement_group import PlacementGroup
+from torchx.schedulers.ray.ray_common import RayActor
+
+_logger: logging.Logger = logging.getLogger(__name__)
+_logger.setLevel(logging.INFO)
+
+
+@contextlib.contextmanager
+def redirect_argv(args: List[str]):  # pyre-ignore[3]
+    _argv = sys.argv[:]
+    sys.argv = args
+    yield
+    sys.argv = _argv
+
+
+@ray.remote
+class CommandActor:  # pragma: no cover
+    def __init__(self, command: str, env: Dict[str, str]) -> None:
+        self.args: List[str] = command.split(" ")
+        self.path: str = self.args[0]
+
+        for k, v in env.items():
+            os.environ[k] = v
+
+    def run_command(self) -> None:
+        spec: Optional[
+            importlib.machinery.ModuleSpec
+        ] = importlib.util.spec_from_file_location("__main__", self.path)
+        if spec:  # pragma: no cover
+            train = importlib.util.module_from_spec(spec)
+            with redirect_argv(self.args):
+                spec.loader.exec_module(train)  # pyre-ignore[16]
+
+
+def load_actor_json(filename: str) -> List[RayActor]:
+    with open(filename) as f:
+        actors: List[RayActor] = []
+        # Yes this is gross but it works
+        actor_dict = json.load(f)
+        actor_dict = json.loads(actor_dict)
+        for actor in actor_dict:
+            actors.append(RayActor(**actor))
+        return actors
+
+
+def create_placement_groups(actors: List[RayActor]) -> List[PlacementGroup]:
+    pgs: List[PlacementGroup] = []
+    for actor in actors:
+        bundle = {"CPU": actor.num_cpus, "GPU": actor.num_gpus}
+        bundles = [bundle] * actor.num_replicas
+        pg = ray.util.placement_group(bundles, strategy="SPREAD")
+        pgs.append(pg)
+
+        _logger.info("Waiting for placement group to start.")
+        ready = pg.wait(timeout_seconds=100)
+
+        if ready:
+            _logger.info("Placement group has started.")
+            _logger.info("Starting remote function")
+        else:  # pragma: no cover
+            raise TimeoutError(
+                "Placement group creation timed out. Make sure "
+                "your cluster either has enough resources or use "
+                "an autoscaling cluster. Current resources "
+                "available: {}, resources requested by the "
+                "placement group: {}".format(ray.available_resources(), pg.bundle_specs)
+            )
+    return pgs
+
+
+def create_command_actors(
+    actors: List[RayActor], pgs: List[PlacementGroup]
+) -> List[CommandActor]:
+    command_actors: List[CommandActor] = []
+    address, port = get_address_and_port()
```

Comment:
what this will return for distributed job, a master node address?

Reply:
From the Ray source code

```python
def get_address_and_port() -> Tuple[str, int]:
    """Returns the IP address and a free port on this node."""

```



```python
master_addr, master_port = worker_group.execute_single(0, get_address_and_port)
```

Question to @amogkam, would it be preferable here to use this like here https://docs.ray.io/en/master/_modules/ray/train/torch.html